educator/src/Dscp/Educator/Web/Auth.hs

```diff
@@ -80,6 +94,130 @@ instance ApiHasArgClass (Auth' auths res) where
+    AuthNotMentioned auth '[] = ()
+    AuthNotMentioned auth (auth ': auths) = TypeError
+        ('Text "Duplicated auth type " ':<>: 'ShowType auth)
+    AuthNotMentioned auth (auth0 ': auths) =
+        AuthNotMentioned auth auths
+
+-- | Whether all types in @auths@ are different.
+type family AllUniqueAuths (auths :: [*]) :: Constraint where
+    AllUniqueAuths '[] = ()
+    AllUniqueAuths (auth ': auths) = (AuthNotMentioned auth auths, AllUniqueAuths auths)
+
+-- | Implementation of 'IsClientAuth' for some @auth@.
+data SomeAuthProvider =
+    forall (auth :: *). Typeable auth =>
+    SomeAuthProvider (Cli.Request -> ClientAuthData auth -> AuthToken)
+
+-- | Multi-version of 'IsClientAuth'.
+class AreClientAuths (auths :: [*]) where
+    provideAuths :: [SomeAuthProvider]
+
+instance AreClientAuths '[] where
+    provideAuths = []
+
+instance (IsClientAuth auth, Typeable auth, AreClientAuths auths) =>
+         AreClientAuths (auth ': auths) where
+    provideAuths = SomeAuthProvider (provideAuth @auth) : provideAuths @auths
+
+-- | Some 'ClientAuthData' complying one of @auths@.
+data ClientAuthDataFrom (auths :: [*]) =
+    forall auth. (KnownClientAuth auth auths, Typeable auth) =>
+    CliAuthData (ClientAuthData auth)
+
+-- | Form an authentication header from provided authentication data.
+encodeAuthData
+    :: forall auths.
+       (Typeable auths, AreClientAuths auths, AllUniqueAuths auths)
+    => Cli.Request -> ClientAuthDataFrom auths -> AuthToken
+encodeAuthData req (CliAuthData authData) =
+    -- The list constructed in the following do-block is not empty due to
+    -- invariants of 'ClientAuthDataFrom'.
+    -- Also, it contains no more than one element since we forced all @auths@ to be
+    -- unique, thus all corresponding 'ClientAuthData's are unique as well.
+    oneOrError $ do
+        SomeAuthProvider authProvider <- provideAuths @auths
+        Just matchingAuthData <- pure (cast authData)
+        pure $ authProvider req matchingAuthData
+
+-- | By built client request provides list of tokens for Authorization header.
+newtype MakeAuthHeaders = MakeAuthHeaders (Cli.Request -> [AuthToken])
+
+-- | Supposedly a wrapper over 'ClientM', it carries 'MakeAuthHeaders' till the moment
+-- a request is fully built and inserts authentication header upon submission.
+newtype AuthClientM m a = AuthClientM
+    { unAuthClientM :: ReaderT MakeAuthHeaders m a
+    } deriving (Functor, Applicative, Monad, MonadIO, MonadThrow, MonadCatch)
+
+runAuthClientM :: MakeAuthHeaders -> AuthClientM m a -> m a
+runAuthClientM ctx action = runReaderT (unAuthClientM action) ctx
+
+-- | Attaches authentication header to given request.
+runAuthRequest :: (Cli.Request -> m a) -> Cli.Request -> AuthClientM m a
+runAuthRequest runReq req =
+    AuthClientM . ReaderT $ \(MakeAuthHeaders makeAuthHeaders) -> do
+        let authHeaders = makeAuthHeaders req
+        let totalAuthHeader = BS.intercalate ", " $ map unAuthHeader authHeaders
```

Comment:
What is the semantics of it? AFAIK, passing a comma-separated list of different authentication tokens in not the valid or standard way of doing authentication; our custom server code does not support this too. So what is the idea behind this?

Reply:
Some people say that it's fine to pass several authentication headers like this
https://stackoverflow.com/questions/29282578/multiple-http-authorization-headers.
Probably it should cause the server to accept the first working authentication method in the list.

Well, it actually doesn't seem canonical, so I'd better allow only one auth credentials.